[diem-vm] Enable parallel execution with on chain config

runtian-zhou opened the pull request:
<!--
Thank you for sending a PR. We appreciate you spending time to help improve the Diem project.

The project is undergoing daily changes. Pull Requests will be reviewed and responded to as time permits.
-->

## Motivation

Implemented an on-chain config to control the parallel execution feature: if the config is set to `Some(static_analysis_result)`, `DiemVM` will enable parallel execution feature and execute transactions in parallel.

This PR has five commits:
1. Make `ReadWriteSet` (the static analysis result generated by `read-write-set`) serializable. 
2. Implement DiemFramework modules to host the config.
3. Implement the on chain config for parallel execution in `diem-types` 
4. Generate boilerplate code in `read-writeset-generater`
5. Implement the logic switching between parallel and sequential execution and test for management script.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/diem/diem/blob/main/CONTRIBUTING.md#pull-requests)?

Yes

## Test Plan

Added a new test case for using management script to enable and disable parallel execution.

libra-action commented:
**LSR dependency change summary**:
```
target packages:
  A move-binary-format 0.1.0 (workspace, path 'language/move-binary-format')
    * features: default
  A read-write-set-types 0.1.0 (workspace, path 'language/tools/read-write-set/types')
    * features: [none]

host packages:
  A variant_count 1.1.0 (direct third-party, crates.io)
    * features: [none]
```
<!-- metadata: {"tag":"lsr-summary"} -->

libra-action commented:
**LSR dependency change summary**:
```
target packages:
  A move-binary-format 0.1.0 (workspace, path 'language/move-binary-format')
    * features: default
  A read-write-set-types 0.1.0 (workspace, path 'language/tools/read-write-set/types')
    * features: [none]

host packages:
  A variant_count 1.1.0 (direct third-party, crates.io)
    * features: [none]
```
<!-- metadata: {"tag":"lsr-summary"} -->

zekun000 commented on diem-move/diem-framework/DPN/releases/artifacts/current/build/DiemCoreFramework/sources/ParallelExecutionConfig.move:
```diff
@@ -0,0 +1,112 @@
+       dr_account: &signer,
+       read_write_inference_result: vector<u8>,
+    ) {
+        DiemTimestamp::assert_operating();
```
do we need this? we might want to initialize in all new genesis

runtian-zhou replied:
Initialize should be done by the `initialize_parallel_config` function IMO

zekun000 replied:
yes, but that can only initialize it to None which means we need to send another transaction after genesis to enable it

zekun000 commented on language/e2e-testsuite/src/tests/parallel_execution.rs:
```diff
@@ -121,3 +122,33 @@ fn rotate_ed25519_key() {
+    // insert a block prologue transaction
+    let (txns_info, transfer_txns) = create_cyclic_transfers(&executor, &accounts, transfer_amount);
+
+    executor.enable_parallel_execution();
```
I think it might be worth testing a malformed config (unable to deserialize) on-chain and make sure it safely fall back

runtian-zhou replied:
Test added!

libra-action commented:
**LSR dependency change summary**:
```
target packages:
  A move-binary-format 0.1.0 (workspace, path 'language/move-binary-format')
    * features: default
  A read-write-set-types 0.1.0 (workspace, path 'language/tools/read-write-set/types')
    * features: [none]

host packages:
  A variant_count 1.1.0 (direct third-party, crates.io)
    * features: [none]
```
<!-- metadata: {"tag":"lsr-summary"} -->

tzakian commented on diem-move/diem-framework/DPN/releases/artifacts/current/build/DiemCoreFramework/sources/ParallelExecutionConfig.move:
```diff
@@ -0,0 +1,112 @@
+        );
+    }
+
+    public fun set_parallel_execution(
```
For parity with `disable_parallel_execution` what do you think of calling this `enable_parallel_execution`?

runtian-zhou replied:
`enable_parallel_execution_with_config`?

tzakian commented on diem-move/diem-framework/DPN/releases/artifacts/current/build/DiemCoreFramework/sources/SystemAdministrationScripts.move:
```diff
@@ -160,4 +161,62 @@ module DiemFramework::SystemAdministrationScripts {
     }
+
+    ///  # Summary
+    /// Updates the ParallelExecutionConfig that is stored on-chain and is used by the parallel executor's read write
```
Nit: back ticks around `ParallelExecutionConfig`

tzakian commented on diem-move/diem-framework/DPN/releases/artifacts/current/build/DiemCoreFramework/sources/SystemAdministrationScripts.move:
```diff
@@ -160,4 +161,62 @@ module DiemFramework::SystemAdministrationScripts {
+    }
+
+    ///  # Summary
+    /// Updates the ParallelExecutionConfig that is stored on-chain to disable parallel execution.  This
```
Nit: backticks here as well

tzakian commented on diem-move/diem-framework/DPN/releases/artifacts/current/build/DiemCoreFramework/sources/SystemAdministrationScripts.move:
```diff
@@ -160,4 +161,62 @@ module DiemFramework::SystemAdministrationScripts {
+    /// | ------          | ------        | -------------                                                              |
+    /// | `account`       | `signer`      | Signer of the sending account. Must be the Diem Root account.              |
+    /// | `sliding_nonce` | `u64`         | The `sliding_nonce` (see: `SlidingNonce`) to be used for this transaction. |
+    /// | `config`        | `vector<u8>`  | The serialized bytes of ReadWriteSet analysis result.                                  |
```
The script doesn't have this argument :)

runtian-zhou replied:
I should probably just get rid of the scripts here as they will be used by admin scripts.

tzakian commented on diem-move/writeset-transaction-generator/templates/disable_parallel_execution.move:
```diff
@@ -0,0 +1,6 @@
+script {
```
I'm wondering if there's a need for the script functions defined in the Move configs commit if the plan is for these to be used (or if only those or these should exist). Are they only there for testing purposes?

runtian-zhou replied:
IMHO they are not only for testing purposes. When we want to enable/disable parallel execution in testing/production environment, we will use those scripts to generate the writeset payload needs to be sent to the network.

runtian-zhou commented on diem-move/diem-vm/src/diem_vm.rs:
```diff
@@ -717,11 +721,25 @@ impl VMExecutor for DiemVM {
-            .into_iter()
-            .map(|(_vm_status, txn_output)| txn_output)
-            .collect())
+        if let Some(read_write_set_analysis) =
```
That is indeed the case! The read-write-set analzyer won't work for writeset transaction and will thus forces the system to go back to sequential mode! As promised in our chat I'll add a test specifically to make sure that the ParallelVM can handle writeset and specifically via sequential execution mode.

runtian-zhou commented:
PR updated! Major changes are:
1. Remove the SystemAdminScripts as the scripts should be used via admin script thus we don't need to expose them for now
2. Renamed `set_parallel_execution` to `enable_parallel_execution_with_config` for duality
3. Added test to make sure writeset transactions are executed sequentially.
4. Added test to make sure sequential execution mode is enabled when having a bad on chain config.

libra-action commented:
**LSR dependency change summary**:
```
target packages:
  A move-binary-format 0.1.0 (workspace, path 'language/move-binary-format')
    * features: default
  A read-write-set-types 0.1.0 (workspace, path 'language/tools/read-write-set/types')
    * features: [none]

host packages:
  A variant_count 1.1.0 (direct third-party, crates.io)
    * features: [none]
```
<!-- metadata: {"tag":"lsr-summary"} -->

tzakian reviewed (approved):
This LGTM!

libra-action commented:
**LSR dependency change summary**:
```
target packages:
  A move-binary-format 0.1.0 (workspace, path 'language/move-binary-format')
    * features: default
  A read-write-set-types 0.1.0 (workspace, path 'language/tools/read-write-set/types')
    * features: [none]

host packages:
  A variant_count 1.1.0 (direct third-party, crates.io)
    * features: [none]
```
<!-- metadata: {"tag":"lsr-summary"} -->

runtian-zhou commented:
Rebased and fixed the tests. Added the initialize config to genesis for on chain config tests to pass.

libra-action commented:
**LSR dependency change summary**:
```
target packages:
  A move-binary-format 0.0.3 (workspace, path 'language/move-binary-format')
    * features: default
  A move-read-write-set-types 0.0.3 (workspace, path 'language/tools/read-write-set/types')
    * features: [none]

host packages:
  A variant_count 1.1.0 (direct third-party, crates.io)
    * features: [none]
```
<!-- metadata: {"tag":"lsr-summary"} -->

libra-action commented:
**LSR dependency change summary**:
```
target packages:
  A move-binary-format 0.0.3 (workspace, path 'language/move-binary-format')
    * features: default
  A move-read-write-set-types 0.0.3 (workspace, path 'language/tools/read-write-set/types')
    * features: [none]

host packages:
  A variant_count 1.1.0 (direct third-party, crates.io)
    * features: [none]
```
<!-- metadata: {"tag":"lsr-summary"} -->

libra-action commented:
**LSR dependency change summary**:
```
target packages:
  A move-binary-format 0.0.3 (workspace, path 'language/move-binary-format')
    * features: default
  A move-read-write-set-types 0.0.3 (workspace, path 'language/tools/read-write-set/types')
    * features: [none]

host packages:
  A variant_count 1.1.0 (direct third-party, crates.io)
    * features: [none]
```
<!-- metadata: {"tag":"lsr-summary"} -->

runtian-zhou commented:
/land
